c2rust-analyze/src/rewrite/span_index.rs

```diff
@@ -0,0 +1,54 @@
+//! Defines the [`SpanIndex`] data structure for looking up MIR statements by span.
+use rustc_span::{BytePos, Span};
+use std::slice;
+
+/// A mapping from [`Span`]s to values of type `T`.  Allows looking up all items whose span is a
+/// subset of some target span.
+pub struct SpanIndex<T> {
+    v: Vec<(Span, T)>,
+}
+
+impl<T> SpanIndex<T> {
+    /// Construct a new [`SpanIndex`], containing the items provided by the iterator `it`.
+    pub fn new(it: impl IntoIterator<Item = (Span, T)>) -> SpanIndex<T> {
+        let mut v = it.into_iter().collect::<Vec<_>>();
+        v.sort_by_key(|(span, _)| span.lo());
+        SpanIndex { v }
+    }
+
+    /// Iterate over items whose spans are entirely contained within `span`.  The order of the
+    /// returned items is unspecified.
+    pub fn lookup<'a>(&'a self, span: Span) -> RangeIter<'a, T> {
+        let data = span.data();
+        let start = self.v.partition_point(|(span, _)| span.lo() < data.lo);
```

Comment:
how do we know that the spans in the partitioned iterator all have a hi <= span.hi (otherwise would they not be fully contained)?

Reply:
nevermind, i see this is handled in the next implementation